File: MATLAB/HFLUX_3.1/hflux_sens.m
function [sens]=hflux_sens(input_data,dis_high_low,T_L_high_low,vts_high_low,shade_high_low)
% HFLUX SENSITIVITY ANALYSIS
% The HFLUX Sensitivity Analysis is an add-on to the HFLUX Stream
% Temperature Solver.
%
% DESCRIPTION:
%   Does a basic sensitivity analysis on HFLUX inputs that are uniform through
%   time and plots results.  Each input parameter is input as it is in HFLUX,
%   and then the program calculates
%   the range of flux using the range of each individual parameter (low and
%   high), while holding all others at their base values.
%
% USAGE:
%   sens = hflux_sens(time_mod,dist_mod,temp_x0_data,temp_t0_data,dim_data,
%   dis_data,dis_high_low,T_L_data,T_L_high_low,met_data,z,bed_data,
%   vts_high_low,sed_type,shade_data,shade_high_low)
%
% INPUT:
%   input_data: see "help hflux" for details
%   dis_high_low: a pair of values that determine the increase and decrease
%   in the discharge value desired by the user.  Enter values in square
%   brackets with the low value first, ie [-0.5 0.5]
%   T_L_high_low: a pair of values that determine the increase and decrease
%   in the temperature of lateral inflows desired by the user.  Enter values
%   in square brackets with the low value first, ie [-2 2]
%   vts_high_low: a pair of values that determine the increase and decrease
%   in the vts coefficient desired by the user.  Enter values
%   in square brackets with the low value first, ie [-0.2 0.2]
%   shade_high_low: a pair of values that determine the increase and decrease
%   in the shading desired by the user.  Enter values
%   in square brackets with the low value first, ie [-0.2 0.2]
%
% EXAMPLE:
%   [sens]=hflux_sens(time_mod,dist_mod,temp_x0_data,temp_t0_data,dim_data,
%   dis_data,[-.01 .01],T_L_data,[-2 2],met_data,[-2 2],
%   shade_data,[-.2 .2])

% Written by AnneMarie Glose, Syracuse University, April 2013
%   Department of Earth Sciences
%   204 Heroy Geology Lab
%   Syracuse, NY  13244
%   Contact: amglose@syr.edu
% Copyright (c) 2013, AnneMarie Glose. All rights reserved.
% Last updated 10/28/2014

disp('Calculating high and low values...')

% Set program option to suppress graphical output
input_data.settings(5)=1;

% Create low and high values for each parameter
dis_low=input_data.dis_data(:,2:end)+dis_high_low(1); %changed to work with more columns
dis_high=input_data.dis_data(:,2:end)+dis_high_low(2); %changed to work with more columns
T_L_low=input_data.T_L_data(:,2)+T_L_high_low(1);
T_L_high=input_data.T_L_data(:,2)+T_L_high_low(2);
vts_low=input_data.shade_data(:,3)+vts_high_low(1);
vts_high=input_data.shade_data(:,3)+vts_high_low(2);
shade_low=input_data.shade_data(:,2)+shade_high_low(1);
shade_high=input_data.shade_data(:,2)+shade_high_low(2);

% Create hflux-ready arrays from the low and high values
dis_data_low = horzcat(input_data.dis_data(:,1),dis_low); %allows for multiple columns of discharge data
dis_data_high = horzcat(input_data.dis_data(:,1),dis_high);
% dis_data_low(:,1)=input_data.dis_data(:,1); %previous code for no changing discharge
% dis_data_low(:,2)=dis_low;
% dis_data_high(:,1)=input_data.dis_data(:,1);
% dis_data_high(:,2)=dis_high;

T_L_data_low(:,1)=input_data.T_L_data(:,1);
T_L_data_low(:,2)=T_L_low;
T_L_data_high(:,1)=input_data.T_L_data(:,1);
T_L_data_high(:,2)=T_L_high;

vts_data_low(:,1)=input_data.shade_data(:,1);
vts_data_low(:,2)=input_data.shade_data(:,2);
vts_data_low(:,3)=vts_low;
vts_data_high(:,1)=input_data.shade_data(:,1);
vts_data_high(:,2)=input_data.shade_data(:,2);
vts_data_high(:,3)=vts_high;

shade_data_low(:,1)=input_data.shade_data(:,1);
shade_data_low(:,2)=shade_low;
shade_data_low(:,3)=input_data.shade_data(:,3);
shade_data_high(:,1)=input_data.shade_data(:,1);
shade_data_high(:,2)=shade_high;
shade_data_high(:,3)=input_data.shade_data(:,3);

% Overwrite values in input data for each case
input_data_base=input_data;
input_data_lowdis=input_data;
input_data_lowdis.dis_data=dis_data_low;
input_data_highdis=input_data;
input_data_highdis.dis_data=dis_data_high;
input_data_lowT_L=input_data;
input_data_lowT_L.T_L_data=T_L_data_low;
input_data_highT_L=input_data;
input_data_highT_L.T_L_data=T_L_data_high;
input_data_lowvts=input_data;
input_data_lowvts.shade_data=vts_data_low;
input_data_highvts=input_data;
input_data_highvts.shade_data=vts_data_high;
input_data_lowshade=input_data;
input_data_lowshade.shade_data=shade_data_low;
input_data_highshade=input_data;
input_data_highshade.shade_data=shade_data_high;

disp('...Done!')
disp('    ')
disp('Running HLUX for the base, high, and low cases...')

% Run hlux.m for middle (base) values, then for high and low values of
% each parameter with other parameters kept at base values
[temp_mod_base, ~, ~, ~]=hflux(input_data_base);
[temp_mod_lowdis, ~, ~, ~]=hflux(input_data_lowdis);
[temp_mod_highdis, ~, ~, ~]=hflux(input_data_highdis);
[temp_mod_lowT_L, ~, ~, ~]=hflux(input_data_lowT_L);
[temp_mod_highT_L, ~, ~, ~]=hflux(input_data_highT_L);
[temp_mod_lowvts, ~, ~, ~]=hflux(input_data_lowvts);
[temp_mod_highvts, ~, ~, ~]=hflux(input_data_highvts);
[temp_mod_lowshade, ~, ~, ~]=hflux(input_data_lowshade);
[temp_mod_highshade, ~, ~, ~]=hflux(input_data_highshade);

disp('...Done!')
disp('    ')
disp('Writing output data...')
% Write outputs from hflux to output structures
base.temp=temp_mod_base;
base.mean=mean(temp_mod_base,2);

lowdis.temp=temp_mod_lowdis;
lowdis.mean=mean(temp_mod_lowdis,2);

highdis.temp=temp_mod_highdis;
highdis.mean=mean(temp_mod_highdis,2);

lowT_L.temp=temp_mod_lowT_L;
lowT_L.mean=mean(temp_mod_lowT_L,2);

highT_L.temp=temp_mod_highT_L;
highT_L.mean=mean(temp_mod_highT_L,2);

lowvts.temp=temp_mod_lowvts;
lowvts.mean=mean(temp_mod_lowvts,2);

highvts.temp=temp_mod_highvts;
highvts.mean=mean(temp_mod_highvts,2);

lowshade.temp=temp_mod_lowshade;
lowshade.mean=mean(temp_mod_lowshade,2);

highshade.temp=temp_mod_highshade;
highshade.mean=mean(temp_mod_highshade,2);

% Write structures to output
sens.dis_l=dis_data_low;
sens.dis_h=dis_data_high;
sens.TL_l=T_L_data_low;
sens.TL_h=T_L_data_high;
sens.vts_l=vts_low;
sens.vts_h=vts_high;
sens.sh_l=shade_data_low;
sens.sh_h=shade_data_high;
sens.base=base;
sens.lowdis=lowdis;
sens.highdis=highdis;
sens.lowT_L=lowT_L;
sens.highT_L=highT_L;
sens.lowvts=lowvts;
sens.highvts=highvts;
sens.lowshade=lowshade;
sens.highshade=highshade;

disp('...Done!')
disp('    ')

% Make sensitivity plots
figure;
subplot(2,2,1)
hold on
plot(input_data.dist_mod,sens.lowdis.mean,'--b','LineWidth',2)
plot(input_data.dist_mod,sens.base.mean,'k','LineWidth',2)
plot(input_data.dist_mod,sens.highdis.mean,'--r','LineWidth',2)
hold off
set(gca,'XLim',[min(input_data.temp_t0_data(:,1)),...
    max(input_data.temp_t0_data(:,1))]) %set x-axis limits to dist range
set(gca,'FontName','Arial')
title('Discharge','FontName', 'Arial','FontSize',14,'FontWeight','Bold'), ...
    xlabel('Distance Downstream (m)','FontName', 'Arial','FontSize',...
    12,'FontWeight','Bold'),...
    ylabel(['Temperature (' setstr(176) 'C)'],'FontName', 'Arial',...
    'FontSize',12,'FontWeight','Bold')
legend('Low','Base','High'); %legend of values (base,high,low)

subplot(2,2,2)
hold on
plot(input_data.dist_mod,sens.lowT_L.mean,'--b','LineWidth',2)
plot(input_data.dist_mod,sens.base.mean,'k','LineWidth',2)
plot(input_data.dist_mod,sens.highT_L.mean,'--r','LineWidth',2)
hold off
set(gca,'XLim',[min(input_data.temp_t0_data(:,1)),...
    max(input_data.temp_t0_data(:,1))]) %set x-axis limits to dist range
set(gca,'FontName','Arial')
title('Groundwater Temperature','FontName', 'Arial','FontSize',14,'FontWeight','Bold'), ...
    xlabel('Distance Downstream (m)','FontName', 'Arial','FontSize',...
    12,'FontWeight','Bold'),...
    ylabel(['Temperature (' setstr(176) 'C)'],'FontName', 'Arial',...
    'FontSize',12,'FontWeight','Bold')
legend('Low','Base','High'); %legend of values (base,high,low)

subplot(2,2,3)
hold on
plot(input_data.dist_mod,sens.lowvts.mean,'--b','LineWidth',2)
plot(input_data.dist_mod,sens.base.mean,'k','LineWidth',2)
plot(input_data.dist_mod,sens.highvts.mean,'--r','LineWidth',2)
hold off
set(gca,'XLim',[min(input_data.temp_t0_data(:,1)),...
    max(input_data.temp_t0_data(:,1))]) %set x-axis limits to dist range
set(gca,'FontName','Arial')
title('View to Sky Coefficient','FontName', 'Arial','FontSize',14,'FontWeight','Bold'), ...
    xlabel('Distance Downstream (m)','FontName', 'Arial','FontSize',...
    12,'FontWeight','Bold'),...
    ylabel(['Temperature (' setstr(176) 'C)'],'FontName', 'Arial',...
    'FontSize',12,'FontWeight','Bold')
legend('Low','Base','High'); %legend of values (base,high,low)

subplot(2,2,4)
hold on
plot(input_data.dist_mod,sens.lowshade.mean,'--b','LineWidth',2)
plot(input_data.dist_mod,sens.base.mean,'k','LineWidth',2)
plot(input_data.dist_mod,sens.highshade.mean,'--r','LineWidth',2)
hold off
set(gca,'XLim',[min(input_data.temp_t0_data(:,1)),...
    max(input_data.temp_t0_data(:,1))]) %set x-axis limits to dist range
set(gca,'FontName','Arial')
title('Shade','FontName', 'Arial','FontSize',14,'FontWeight','Bold'), ...
    xlabel('Distance Downstream (m)','FontName', 'Arial','FontSize',...
    12,'FontWeight','Bold'),...
    ylabel(['Temperature (' setstr(176) 'C)'],'FontName', 'Arial',...
    'FontSize',12,'FontWeight','Bold')
legend('Low','Base','High'); %legend of values (base,high,low)

% Calculate total percent change in stream temperature
change=[mean(mean((sens.lowdis.temp))-mean(mean(sens.base.temp))),...
    (mean(mean(sens.highdis.temp))-mean(mean(sens.base.temp)));...
    (mean(mean(sens.lowT_L.temp))-mean(mean(sens.base.temp))),...
    (mean(mean(sens.highT_L.temp))-mean(mean(sens.base.temp)));...
    (mean(mean(sens.lowvts.temp))-mean(mean(sens.base.temp))),...
    (mean(mean(sens.highvts.temp))-mean(mean(sens.base.temp)));...
    (mean(mean(sens.lowshade.temp))-mean(mean(sens.base.temp))),...
    (mean(mean(sens.highshade.temp))-mean(mean(sens.base.temp)))];
figure;
colormap(hot);
bar(change,'grouped')
labels={'Discharge';'GW Temp';'VTS';'Shade'};
title('Change in Average Stream Temperature With Changing Variables',...
    'FontName','Arial','FontSize',14,'FontWeight','Bold');
ylabel(['Change (' char(176) 'C)'],'FontName','Arial','FontSize',12,'FontWeight','Bold');
xlabel('Adjusted Variable','FontName','Arial','FontSize',12,'FontWeight','Bold');
set(gca,'xticklabel',labels,'FontName','Arial','FontSize',12);
legend('Decrease Value','Increase Value','Location','Best');


sens.change=change;
end %function

% Copyright (c) 2013, AnneMarie Glose.
% All rights reserved.
% 
% Redistribution and use in source and binary forms, with or without
% modification, are permitted provided that the following conditions are
% met:
% (1) Redistributions of source code must retain the above copyright
% notice, this list of conditions and the following disclaimer.
% (2) Redistributions in binary form must reproduce the above copyright
% notice, this list of conditions and the following disclaimer in the
% documentation and/or other materials provided with the distribution.
% 
% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
% IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
% THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
% PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
% CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
% EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
% PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
% PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
% LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
% NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
% SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

File: Python/src/Heat_Flux/hflux_sens.py
"""
Author: Michael Scoleri, Violet Shi
Date: 10-02-23
File: hflux_sens.py
Functionality: Implementation of hflux_sens.m
"""

import sys
from concurrent.futures import ProcessPoolExecutor
import multiprocessing
from Python.src.Core.heat_flux import HeatFlux
from Python.src.Plotting.plotting_class import Plotting
import numpy as np
import matplotlib.pyplot as plt


class HfluxSens:
    def __init__(self, root_dir):
        self.plc = Plotting()
        self.root_dir = root_dir
        sys.path.append(root_dir)

    def class_assistant(self):
        """
        Print out class details.
        """
        help(type(self))

    @staticmethod
    def heat_flux_wrapper(data_table):
        """
        Wrapper for calling heat flux.

        Args:
            data_table (Data table class object): Class object containing ndarrays of data.

        Returns:
            Result from heat flux calculations.
        """
        return HeatFlux(data_table).crank_nicolson_method()

    @staticmethod
    def multithreading_call(output_suppression, input_data_list, base_result):
        """
        Implements Multi-threading for calls to hflux.py

        Args:
            input_data_list ([dict]): list of dictionaries to be used as the parameter for hflux.py

        Return:
            results ({ndarray, {ndarrarys}, {ndarrays}})
        """
        results = [base_result]

        # Launching Parallel Tasks - https://docs.python.org/3/library/concurrent.futures.html
        if not output_suppression:
            print()
            print("Beginning Multi-threaded calls to hflux...")

        cpu_count = multiprocessing.cpu_count()
        with ProcessPoolExecutor(max_workers=cpu_count // 4) as executor:
            for result, _, _, _ in executor.map(
                HfluxSens.heat_flux_wrapper, input_data_list
            ):
                results.append(result)
            executor.shutdown()

        if not output_suppression:
            print("...closed multithreading executer...")

        return results

    @staticmethod
    def singlethread_call(output_suppression, input_data_list, base_result):
        """
        Calls heat_flux.py in a single thread

        Args:
            input_data_list ([dict]): list of dictionaries to be used as the parameter for hflux.py

        Return:
            results ({ndarray, {ndarrarys}, {ndarrays}})
        """
        if not output_suppression:
            print()
            print("Beginning calls to hflux...")

        results = [base_result]
        for result, _, _, _ in map(HfluxSens.heat_flux_wrapper, input_data_list):
            results.append(result)
            if not output_suppression:
                print("Finished a call!")

        if not output_suppression:
            print()
            print("Sensitivity values computed.")

        return results

    def hflux_sens(
        self,
        data_table,
        dis_high_low,
        t_l_high_low,
        vts_high_low,
        shade_high_low,
    ):
        """
        Creates and returns a dictionary with new high and low values.

        Args:
            data_table (data table class object): Class object containing ndarrays of data.
            dis_high_low (ndarray): Ndarray containing distance high and low data.
            t_l_high_low (ndarray): Ndarray containing time high and low data.
            vts_high_low (ndarray): Ndarray containing vts high and low data.
            shade_high_low (ndarray): Ndarray containg shade high and low data.

        Returns:
            High_low_values ({ndarray}): Dictionary containing high and low values.
        """
        output_suppression = data_table.output_suppression
        if not output_suppression:
            print("Calculating high and low values...")

        data_table.output_suppression = True

        high_low_values = {}
        # Create low and high values for each parameter.
        high_low_values["dis_data_1"] = data_table.discharge.transpose()
        high_low_values["dis_low"] = high_low_values["dis_data_1"] + dis_high_low[0]
        high_low_values["dis_high"] = high_low_values["dis_data_1"] + dis_high_low[1]

        high_low_values["t_l_data_1"] = data_table.t_L
        high_low_values["t_l_low"] = high_low_values["t_l_data_1"] + t_l_high_low[0]
        high_low_values["t_l_high"] = high_low_values["t_l_data_1"] + t_l_high_low[1]

        high_low_values["vts"] = data_table.vts
        high_low_values["vts_low"] = high_low_values["vts"] + vts_high_low[0]
        high_low_values["vts_high"] = high_low_values["vts"] + vts_high_low[1]

        high_low_values["shade_1"] = data_table.shade
        high_low_values["shade_low"] = high_low_values["shade_1"] + shade_high_low[0]
        high_low_values["shade_high"] = high_low_values["shade_1"] + shade_high_low[1]

        # Create hflux-ready arrays from the low and high values
        # Use hstack to concatenate along the 2nd axis.
        # Use newaxis to index arrays.
        # np.hstack - https://numpy.org/devdocs/reference/generated/numpy.hstack.html#numpy-hstack
        # np.newaxis - https://numpy.org/devdocs/reference/constants.html#numpy.newaxis
        high_low_values["dis_data_0"] = data_table.dist_dis
        high_low_values["dis_data_low"] = np.hstack(
            (high_low_values["dis_data_0"][:, np.newaxis], high_low_values["dis_low"])
        ).transpose()
        high_low_values["dis_data_high"] = np.hstack(
            (high_low_values["dis_data_0"][:, np.newaxis], high_low_values["dis_high"])
        ).transpose()

        # Set t_l_data_low and t_l_data_high values.
        high_low_values["t_l_data_0"] = data_table.dist_T_L
        high_low_values["t_l_data_low"] = np.array(
            [high_low_values["t_l_data_0"], high_low_values["t_l_low"]]
        )
        high_low_values["t_l_data_high"] = np.array(
            [high_low_values["t_l_data_0"], high_low_values["t_l_high"]]
        )

        # Set vts_data_low and vts_data_high values.
        high_low_values["shade_0"] = data_table.dist_shade
        high_low_values["vts_data_low"] = np.array(
            [
                high_low_values["shade_0"],
                high_low_values["shade_1"],
                high_low_values["vts_low"],
            ]
        )
        high_low_values["vts_data_high"] = np.array(
            [
                high_low_values["shade_0"],
                high_low_values["shade_1"],
                high_low_values["vts_high"],
            ]
        )

        # Initalize shade_data_low and shade_data_high with zeros.
        high_low_values["shade_2"] = data_table.vts
        high_low_values["shade_data_low"] = np.array(
            [
                high_low_values["shade_0"],
                high_low_values["shade_low"],
                high_low_values["shade_2"],
            ]
        )
        high_low_values["shade_data_high"] = np.array(
            [
                high_low_values["shade_0"],
                high_low_values["shade_high"],
                high_low_values["shade_2"],
            ]
        )

        if not output_suppression:
            print("...done!\n")
            print("Running HLUX for the base, high, and low cases...")

        # Create multiple copies of data_table for and modifying specifc keys.
        high_low_values["input_data_lowdis"] = data_table.modify_data_table(
            "dis_data", high_low_values["dis_data_low"]
        )
        high_low_values["input_data_highdis"] = data_table.modify_data_table(
            "dis_data", high_low_values["dis_data_high"]
        )
        high_low_values["input_data_lowT_L"] = data_table.modify_data_table(
            "t_l_data", high_low_values["t_l_data_low"]
        )
        high_low_values["input_data_highT_L"] = data_table.modify_data_table(
            "t_l_data", high_low_values["t_l_data_high"]
        )
        high_low_values["input_data_lowvts"] = data_table.modify_data_table(
            "shade_data", high_low_values["vts_data_low"]
        )
        high_low_values["input_data_highvts"] = data_table.modify_data_table(
            "shade_data", high_low_values["vts_data_high"]
        )
        high_low_values["input_data_lowshade"] = data_table.modify_data_table(
            "shade_data", high_low_values["shade_data_low"]
        )
        high_low_values["input_data_highshade"] = data_table.modify_data_table(
            "shade_data", high_low_values["shade_data_high"]
        )
        return high_low_values

    def create_new_results(
        self, base_result, high_low_dict, output_suppression, multithread=True
    ):
        """
        Creates and returns a new sensitiviity dictionary

        Args:
            Base_result (ndarray): ndarray containing the temperature modulus data.
            high_low_dict ({ndarray}): Dictionary containing high low values calculated from hflux_sens function.

        Returns:
            Sens ({ndarary}): Dictionary containing sensitivity values.
        """
        new_data_list = [
            high_low_dict["input_data_lowdis"],
            high_low_dict["input_data_highdis"],
            high_low_dict["input_data_lowT_L"],
            high_low_dict["input_data_highT_L"],
            high_low_dict["input_data_lowvts"],
            high_low_dict["input_data_highvts"],
            high_low_dict["input_data_lowshade"],
            high_low_dict["input_data_highshade"],
        ]

        if multithread:
            results = self.multithreading_call(
                output_suppression,
                input_data_list=new_data_list,
                base_result=base_result,
            )
        else:
            results = self.singlethread_call(
                output_suppression,
                input_data_list=new_data_list,
                base_result=base_result,
            )

        temp_mod_base = results[0]
        temp_mod_lowdis = results[1]
        temp_mod_highdis = results[2]
        temp_mod_lowT_L = results[3]
        temp_mod_highT_L = results[4]
        temp_mod_lowvts = results[5]
        temp_mod_highvts = results[6]
        temp_mod_lowshade = results[7]
        temp_mod_highshade = results[8]

        if not output_suppression:
            print("...Done!\n")
            print("Writing output data...\n")

        # Store outputs from hflux to dictionaries.
        base = {"temp": temp_mod_base, "mean": np.mean(temp_mod_base, axis=1)}
        lowdis = {"temp": temp_mod_lowdis, "mean": np.mean(temp_mod_lowdis, axis=1)}
        highdis = {"temp": temp_mod_highdis, "mean": np.mean(temp_mod_highdis, axis=1)}
        lowT_L = {"temp": temp_mod_lowT_L, "mean": np.mean(temp_mod_lowT_L, axis=1)}
        highT_L = {"temp": temp_mod_highT_L, "mean": np.mean(temp_mod_highT_L, axis=1)}
        lowvts = {"temp": temp_mod_lowvts, "mean": np.mean(temp_mod_lowvts, axis=1)}
        highvts = {"temp": temp_mod_highvts, "mean": np.mean(temp_mod_highvts, axis=1)}
        lowshade = {
            "temp": temp_mod_lowshade,
            "mean": np.mean(temp_mod_lowshade, axis=1),
        }
        highshade = {
            "temp": temp_mod_highshade,
            "mean": np.mean(temp_mod_highshade, axis=1),
        }

        # Store structures in dictionary
        sens = {
            "dis_l": high_low_dict["dis_data_low"],
            "dis_h": high_low_dict["dis_data_high"],
            "TL_l": high_low_dict["t_l_data_low"],
            "TL_h": high_low_dict["t_l_data_high"],
            "vts_l": high_low_dict["vts_low"],
            "vts_h": high_low_dict["vts_high"],
            "sh_l": high_low_dict["shade_data_low"],
            "sh_h": high_low_dict["shade_data_high"],
            "base": base,
            "lowdis": lowdis,
            "highdis": highdis,
            "lowT_L": lowT_L,
            "highT_L": highT_L,
            "lowvts": lowvts,
            "highvts": highvts,
            "lowshade": lowshade,
            "highshade": highshade,
        }
        return sens

    def calculate_change(self, sens):
        change = np.array(
            [
                [
                    np.mean(sens["lowdis"]["temp"]) - np.mean(sens["base"]["temp"]),
                    np.mean(sens["highdis"]["temp"]) - np.mean(sens["base"]["temp"]),
                ],
                [
                    np.mean(sens["lowT_L"]["temp"]) - np.mean(sens["base"]["temp"]),
                    np.mean(sens["highT_L"]["temp"]) - np.mean(sens["base"]["temp"]),
                ],
                [
                    np.mean(sens["lowvts"]["temp"]) - np.mean(sens["base"]["temp"]),
                    np.mean(sens["highvts"]["temp"]) - np.mean(sens["base"]["temp"]),
                ],
                [
                    np.mean(sens["lowshade"]["temp"]) - np.mean(sens["base"]["temp"]),
                    np.mean(sens["highshade"]["temp"]) - np.mean(sens["base"]["temp"]),
                ],
            ]
        )
        return change

    # Reshape used to align plotting structures - https://numpy.org/doc/stable/reference/generated/numpy.reshape.html
    def make_sens_plots(self, data_table, sens, return_graphs=False):
        """
        Creates and saves plots for sensitiity calculations

        Args:
            Data_table (Data table class object): Class object containing ndarrays of data.
            Sens ({ndarrays}): Sensitivity Dictionary calculted from create_new_results function.

        Returns:
            None
        """
        fig, ax = plt.subplots(2, 2)

        self.plc.make_three_line_plot(
            x=data_table.dist_mod,
            low_y=sens["lowdis"]["mean"],
            base_y=sens["base"]["mean"],
            high_y=sens["highdis"]["mean"],
            title="Discharge",
            xlabel="Distance Downstream(m)",
            ylabel="Temperature (°C)",
            ax=ax[0, 0],
        )
        self.plc.make_three_line_plot(
            x=data_table.dist_mod,
            low_y=sens["lowT_L"]["mean"],
            base_y=sens["base"]["mean"],
            high_y=sens["highT_L"]["mean"],
            title="Groundwater Temperature",
            xlabel="Distance Downstream(m)",
            ylabel="Temperature (°C)",
            ax=ax[0, 1],
        )
        self.plc.make_three_line_plot(
            x=data_table.dist_mod,
            low_y=sens["lowvts"]["mean"],
            base_y=sens["base"]["mean"],
            high_y=sens["highvts"]["mean"],
            title="View to Sky Coefficient",
            xlabel="Distance Downstream(m)",
            ylabel="Temperature (°C)",
            ax=ax[1, 0],
        )
        self.plc.make_three_line_plot(
            x=data_table.dist_mod,
            low_y=sens["lowshade"]["mean"],
            base_y=sens["base"]["mean"],
            high_y=sens["highshade"]["mean"],
            title="Shade",
            xlabel="Distance Downstream(m)",
            ylabel="Temperature (°C)",
            ax=ax[1, 1],
        )
        plt.tight_layout()

        change = self.calculate_change(sens)
        fig2 = self.plc.make_bar_charts(change)

        self.plc.save_plots(fig, fig2, path="hflux_sens")

        if return_graphs:
            return fig, fig2
